README.md
<center>
    <img src="colorise-logo.png" />
    <h3 align="center">v1.0.0</h3>
    <div align="center">
        <a href="https://travis-ci.org/MisanthropicBit/colorise">
            <img src="https://travis-ci.org/MisanthropicBit/colorise.svg?branch=rgb_256_exts" />
        </a>
        <a href="/LICENSE">
            <img src="https://img.shields.io/github/license/MisanthropicBit/colorise.svg" />
        </a>
        <a href="https://pypi.org/project/colorise/">
            <img src="https://img.shields.io/pypi/v/colorise.svg" />
        </a>
        <img src="https://img.shields.io/pypi/pyversions/colorise.svg" />
    </div>
</center>

---

`colorise` provides easy cross-platform text coloring for terminals/consoles.

## Features

* Supports 8, 16, 88, 256 colors and true-color.
* Colors can be specified by name, index, hexadecimal, HLS, HSV or RGB formats.

```python
>>> colorise.cprint('Hello', fg='red')
>>> colorise.cprint('Hello', fg=201)
>>> colorise.cprint('Hello', fg='#a696ff')
>>> colorise.cprint('Hello', fg='0xa696ff')
>>> colorise.cprint('Hello', fg='hls(0.6923;0.7960;1.0)')
>>> colorise.cprint('Hello', fg='hsv(249;41;100)')
>>> colorise.cprint('Hello', fg='rgb(167;151;255)')
```

* Custom color format akin to Python 3.0 [string formatting](https://docs.python.org/3.7/library/stdtypes.html#str.format).

```python
>>> colorise.fprint('{fg=red}Hello {bg=blue}world!')
>>> colorise.fprint('{fg=red}Hello {bg=blue}world!', autoreset=True)
```

* Automatically find the closest color based on the terminal's
  capabilities. Below is sprite of a [familiar plumber](/examples/mario.py).
  Pixels are specified as RGB so `colorise` automatically approximates colors
  for 256 and 16 color indices in the two right-most images.

<div align="center">
    <img src="/screenshots/mario-true-color.png" width="150" />
    <img src="/screenshots/mario-256-color.png" width="150" />
    <img src="/screenshots/mario-16-color.png" width="150" />
</div>
<div align="center">
    <i>From left to right: True-color, 256 color and 16 color.</i>
</div>
<br />

* Useful functions like [`highlight`](/examples/highlighting.py) that highlights
  individual characters in a string given a list of indices.

```python
>>> colorise.highlight('Hello world', indices=[0, 2, 3, 7, 9], attributes=[Attr.Italic])
```

---

`colorise` has been tested with Python 2.7, 3.2 and 3.3, 3.4, 3.5, 3.6, 3.7 and
pypy, as well as the following terminals:

* [iTerm 3.2.9](https://iterm2.com/)
* Terminal.app 2.9.4 (421.1.1)
* The default Windows console (Windows 8.1)
* [ConEmu](https://conemu.github.io/) vX.X.X (Windows 8.1)

Something not working as expected with your terminal? Please report an
[issue](https://github.com/MisanthropicBit/colorise/issues) or submit a [pull
request](https://github.com/MisanthropicBit/colorise/pulls) using the
[contribution guidelines](/CONTRIBUTING.md).

## Installation

It is recommended to install `colorise` via [`pip`](https://pip.pypa.io/en/latest/).

```bash
pip install colorise
```

## Usage

To print colored text, you can use the `cprint` function.

```python
>>> colorise.cprint('This has blue text and a green background', fg='blue', bg='green')
```

Additionally, you can specify colors via a format string passed to `fprint`.

```python
>>> colorise.fprint('{fg=blue,bg=green}This has blue text and a green background')
```

Text attributes are also supported.

```python
>>> from colorise.attributes import Attr
>>> colorise.cprint('This has blue text and a green background', fg='blue', bg='green', attributes=[Attr.Bold])
>>> colorise.fprint('{fg=blue,bg=green,bold}This has blue text and a green background')
```

You can find additional examples in the [`examples`](/examples) folder.

## Implementation notes

On Linux and Unix there are no 'dark' themed backgrounds (perhaps "bolded" or
"bright" colors are a better name, but do not truly represnt darkened colors),
so calling

```bash
>>> colorise.cprint("Isn't this wrong?", bg='darkred')
```

will not necessarily give the expected result. This is due to the many different
termnial/console types, making it is virtually impossible to correctly map color
names to their actual colors. Additionally, some consoles, such as iTerm and the
Windows standard console, lets you redefine colors, so that using logical
colornames is no guarantee for consistency. ``colorise`` assumes the following
available colors on all systems:

* Black
* Red
* Green
* Yellow
* Blue
* Magenta
* Cyan
* White

It is therefore best to stick to these systems colors which are more likely to
be present and "correct", like 'red', 'blue' and 'green'.


CHANGES.md
# Changes

## 1.0.0

**Major update**

- Reworked entire library
- Better documentation and comments
- Support for RGB, HSV, HLS and hexadecimal color formats
- Support for 88 and 256 color indices
- Support for virtual terminal processing on Windows
- Changed parser to use Python's str.format syntax, e.g. '<fg=red>' -> '{fg=red}''
- Removed formatcolor and formatbyindex functions
- Removed ColorManager classes since no state needs to be stored
- Better detection of terminal color capabilities
- If e.g. a true-color color is specified and the terminal does not support it,
  colorise will automatically find the color on your system that most closely
  matches the desired color

## 0.1.4

- Fixed a bug on nix platforms that caused background colors to break

## 0.1.3

- Fixed a bug where passing a string without any color formatting would print
  the empty string

## 0.1.2

- Fixed a bug in ``nix/ColorManager.py`` which caused ``set_color`` to
  malfunction

## 0.1.1

- Fixed a bug where putting a ``:`` or escaped ``>`` or ``<`` just before or
  after some color formatted text would raise a ``ColorSyntaxError``

## 0.1.0

- Initial version


colorise/nix/cluts.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""Nix color look-up tables (CLUTs) and functions."""

from colorise.color_tools import closest_color
from colorise.terminal import terminal_name
import collections
import os
import curses


_COLOR_ESCAPE_CODE = '\033['
_COLOR_PREFIX_16 = _COLOR_ESCAPE_CODE + '{0}m'
_COLOR_PREFIX_FG_88 = _COLOR_ESCAPE_CODE + '38;5;{0}m'
_COLOR_PREFIX_BG_88 = _COLOR_ESCAPE_CODE + '48;5;{0}m'
_COLOR_PREFIX_FG_256 = _COLOR_PREFIX_FG_88
_COLOR_PREFIX_BG_256 = _COLOR_PREFIX_BG_88
_COLOR_PREFIX_FG_TRUE_COLOR = _COLOR_ESCAPE_CODE + '38;2;{0}m'
_COLOR_PREFIX_BG_TRUE_COLOR = _COLOR_ESCAPE_CODE + '48;2;{0}m'

_prefix_map = {
        #       Foreground prefix            Background prefix
        8:     (_COLOR_PREFIX_16,            _COLOR_PREFIX_16),
        16:    (_COLOR_PREFIX_16,            _COLOR_PREFIX_16),
        88:    (_COLOR_PREFIX_FG_88,         _COLOR_PREFIX_BG_88),
        256:   (_COLOR_PREFIX_FG_256,        _COLOR_PREFIX_BG_256),
        2**24: (_COLOR_PREFIX_FG_TRUE_COLOR, _COLOR_PREFIX_BG_TRUE_COLOR)
    }

# System base colors that are assumed to be present for 8 color terminals
_NIX_SYSTEM_COLORS = {
    30: (0x00, 0x00, 0x00),  # black
    31: (0xff, 0x00, 0x00),  # red
    32: (0x00, 0xff, 0x00),  # green
    33: (0xff, 0xff, 0x00),  # yellow
    34: (0x00, 0x00, 0xff),  # blue
    35: (0xff, 0x00, 0xff),  # purple
    36: (0x00, 0xff, 0xff),  # cyan
    37: (0xff, 0xff, 0xff),  # white
    # 90: (0x64, 0x64, 0x64),  # gray
}

# System color names as given by colorise
_NIX_SYSTEM_COLOR_NAMES = {
    'black':       30,
    'red':         31,
    'green':       32,
    'yellow':      33,
    'blue':        34,
    'purple':      35,
    'cyan':        36,
    'gray':        37,
    'lightblack':  90,
    'lightred':    91,
    'lightgreen':  92,
    'lightyellow': 93,
    'lightblue':   94,
    'lightpurple': 95,
    'lightcyan':   96,
    'lightgray':   97,
}

# Alias for British english gray and alias for purple
_NIX_SYSTEM_COLOR_NAMES['grey'] = 37
_NIX_SYSTEM_COLOR_NAMES['lightgrey'] = 97
_NIX_SYSTEM_COLOR_NAMES['magenta'] = 35

# xterm 88-color look-up table (based on 88colres.h)
_XTERM_CLUT_88_STEPS = [0x00, 0x8b, 0xcd, 0xff]
_XTERM_CLUT_88_GRAYSCALE = [46, 92, 113, 139, 162, 185, 208, 231]
_XTERM_CLUT_88 = collections.OrderedDict(_NIX_SYSTEM_COLORS)

_XTERM_CLUT_88.update(zip(
        range(16, 88),
        [(r, g, b)
         for r in _XTERM_CLUT_88_STEPS
         for g in _XTERM_CLUT_88_STEPS
         for b in _XTERM_CLUT_88_STEPS])
    )

_XTERM_CLUT_88.update(zip(
        range(80, 89),
        [(g, g, g) for g in _XTERM_CLUT_88_GRAYSCALE])
    )

# xterm 256-color look-up table
_XTERM_CLUT_256_STEPS = [0x00, 0x5f, 0x87, 0xaf, 0xd7, 0xff]
_XTERM_CLUT_256_GRAYSCALE = [8 + 10 * i for i in range(24)]
_XTERM_CLUT_256 = collections.OrderedDict(_NIX_SYSTEM_COLORS)

_XTERM_CLUT_256.update(zip(
        range(16, 232),
        [(r, g, b)
         for r in _XTERM_CLUT_256_STEPS
         for g in _XTERM_CLUT_256_STEPS
         for b in _XTERM_CLUT_256_STEPS])
    )

_XTERM_CLUT_256.update(zip(
        range(232, 256),
        [(g, g, g) for g in _XTERM_CLUT_256_GRAYSCALE])
    )


def get_prefix(color_count, bg):
    """Get the color code prefix corresponding to the supported color count."""
    return _prefix_map[color_count][int(bg)]


def get_clut(color_count):
    """Return the appropriate color look-up table."""
    if terminal_name() == 'iTerm.app':
        if color_count == 88:
            # This appears to simply use the 256 color table for 88 color
            # indices
            return _XTERM_CLUT_256

    return {
            8:  _NIX_SYSTEM_COLORS,
            16: _NIX_SYSTEM_COLORS,
            88: _XTERM_CLUT_88,
            256: _XTERM_CLUT_256,
        }.get(color_count, None)


# NOTE: Can use TERM_PROGRAM/TERM_PROGRAM_VERSION to detect terminals
def num_colors():
    """Attempt to get the number of colors supported by the terminal."""
    # iTerm supports true-color from version 3 onward, earlier versions
    # supported 256 colors
    if terminal_name() == 'iTerm.app':
        version = os.environ.get('TERM_PROGRAM_VERSION', '')

        if version and int(version.split('.')[0]) > 2:
            return 2**24
        else:
            return 256

    # If all else fails, use curses
    color_count = 0

    try:
        curses.setupterm()
        color_count = curses.tigetnum('colors')
    except curses.error:
        pass

    if color_count <= 0:
        # Failed to get color count from curses, default to 16 colors
        return 16

    return color_count


def can_redefine_colors():
    """Return whether the terminal allows redefinition of colors."""
    return False


def color_from_name(name, color_count, bg):
    """Return the color value and color count for a given color name."""
    if name not in _NIX_SYSTEM_COLOR_NAMES:
        raise ValueError("Unknown color name '{0}'".format(name))

    # We just return the 16 color prefix and code. Assumes this can can be done
    # in terminals with more capabilities
    return _COLOR_PREFIX_16, _NIX_SYSTEM_COLOR_NAMES[name] + 10 * int(bg)


def color_from_index(idx, color_count, bg):
    """Return the color value and color count for a given color index."""
    if idx < 0 or idx > 255:
        raise ValueError('Color index must be in range 0-256 inclusive')

    if color_count > 88:
        # NOTE: Assume we can use 256 color prefixes if we have true-color
        return (_COLOR_PREFIX_BG_256 if bg else _COLOR_PREFIX_FG_256), idx
    elif color_count == 88:
        prefix = _COLOR_PREFIX_BG_88 if bg else _COLOR_PREFIX_FG_88

        if terminal_name() == 'iTerm.app':
            return prefix, idx

        if idx <= 88:
            return prefix, idx
        else:
            # Approximate > 88 color index with 256 color clut
            return prefix,\
                closest_color(_XTERM_CLUT_256[idx], _XTERM_CLUT_88)
    else:
        if idx <= 16:
            return _COLOR_PREFIX_16, idx + 10 * int(bg)
        else:
            # Approximate > 16 color index with 256 color clut
            key = closest_color(_XTERM_CLUT_256[idx], _NIX_SYSTEM_COLORS)

            return _COLOR_PREFIX_16, key + 10 * int(bg)
